// @bankr/cli/dist/commands/config.js
import { DEFAULT_API_URL, DEFAULT_LLM_URL, getConfigPath, readConfig, writeConfig, } from "../lib/config.js";
import * as output from "../lib/output.js";
const ALLOWED_KEYS = ["apiKey", "apiUrl", "llmKey", "llmUrl"];
const KEY_LABELS = {
    apiKey: "Bankr API Key",
    apiUrl: "Bankr API URL",
    llmKey: "Bankr LLM Key",
    llmUrl: "Bankr LLM Gateway URL",
};
function isValidKey(key) {
    return ALLOWED_KEYS.includes(key);
}
function requireValidKey(key) {
    if (!isValidKey(key)) {
        output.error(`Unknown config key "${key}". Valid keys: ${ALLOWED_KEYS.join(", ")}`);
        process.exit(1);
    }
}
function displayValue(key, value) {
    return key === "apiKey" || key === "llmKey"
        ? output.maskApiKey(value)
        : value;
}
function fallbackValue(key) {
    if (key === "apiUrl")
        return `(default: ${DEFAULT_API_URL})`;
    if (key === "llmUrl")
        return `(default: ${DEFAULT_LLM_URL})`;
    return "(not set)";
}
export async function configGetCommand(key) {
    const config = readConfig();
    if (key) {
        requireValidKey(key);
        const value = config[key];
        if (value !== undefined) {
            console.log(displayValue(key, value));
        }
        else {
            output.dim(fallbackValue(key));
        }
        return;
    }
    output.label("Config file", getConfigPath());
    console.log();
    for (const k of ALLOWED_KEYS) {
        const value = config[k];
        const display = value !== undefined ? displayValue(k, value) : fallbackValue(k);
        output.label(KEY_LABELS[k], display);
    }
}
export async function configSetCommand(key, value) {
    requireValidKey(key);
    const config = readConfig();
    config[key] = value;
    writeConfig(config);
    output.success(`Set ${KEY_LABELS[key]} = ${displayValue(key, value)}`);
}
