// @bankr/cli/dist/commands/submit.js
import { submit } from "../lib/api.js";
import * as output from "../lib/output.js";
function displayResult(result) {
    if (!result.success) {
        output.error(result.error || "Submission failed");
        if (result.transactionHash) {
            output.keyValue("Transaction Hash", result.transactionHash);
        }
        process.exit(1);
    }
    output.success("Transaction submitted");
    console.log();
    output.keyValue("Transaction Hash", result.transactionHash || "");
    output.keyValue("Status", result.status || "unknown");
    output.keyValue("Signer", result.signer);
    output.keyValue("Chain ID", result.chainId.toString());
    if (result.blockNumber) {
        output.keyValue("Block Number", result.blockNumber);
    }
    if (result.gasUsed) {
        output.keyValue("Gas Used", result.gasUsed);
    }
}
export async function submitCommand(opts) {
    // Validate required fields
    if (!opts.to) {
        output.error("--to is required");
        process.exit(1);
    }
    if (!opts.chainId) {
        output.error("--chain-id is required");
        process.exit(1);
    }
    const chainId = parseInt(opts.chainId, 10);
    if (isNaN(chainId)) {
        output.error("--chain-id must be a number");
        process.exit(1);
    }
    let nonce;
    if (opts.nonce) {
        nonce = parseInt(opts.nonce, 10);
        if (isNaN(nonce)) {
            output.error("--nonce must be a number");
            process.exit(1);
        }
    }
    const request = {
        transaction: {
            to: opts.to,
            chainId,
            value: opts.value,
            data: opts.data,
            gas: opts.gas,
            gasPrice: opts.gasPrice,
            maxFeePerGas: opts.maxFeePerGas,
            maxPriorityFeePerGas: opts.maxPriorityFeePerGas,
            nonce,
        },
        description: opts.description,
        waitForConfirmation: !opts.noWait,
    };
    // Execute submission with spinner
    const spinner = output.spinner(opts.noWait ? "Submitting transaction..." : "Submitting and waiting for confirmation...");
    try {
        const result = await submit(request);
        spinner.stop();
        displayResult(result);
    }
    catch (err) {
        spinner.stop();
        output.error(err.message);
        process.exit(1);
    }
}
// Alternative: submit from JSON
export async function submitJsonCommand(json, opts) {
    let transaction;
    try {
        transaction = JSON.parse(json);
    }
    catch {
        output.error("Invalid JSON. Expected: {to, chainId, value?, data?}");
        process.exit(1);
    }
    if (!transaction || typeof transaction !== "object" || Array.isArray(transaction)) {
        output.error("Invalid JSON. Expected object: {to, chainId, value?, data?}");
        process.exit(1);
    }
    // Validate required fields
    if (!transaction.to) {
        output.error("Transaction must have 'to' field");
        process.exit(1);
    }
    if (transaction.chainId === undefined) {
        output.error("Transaction must have 'chainId' field");
        process.exit(1);
    }
    if (typeof transaction.chainId !== "number" || isNaN(transaction.chainId)) {
        output.error("Transaction 'chainId' must be a number");
        process.exit(1);
    }
    if (transaction.nonce !== undefined && (typeof transaction.nonce !== "number" || isNaN(transaction.nonce))) {
        output.error("Transaction 'nonce' must be a number");
        process.exit(1);
    }
    const request = {
        transaction,
        description: opts.description,
        waitForConfirmation: !opts.noWait,
    };
    // Execute submission with spinner
    const spinner = output.spinner(opts.noWait ? "Submitting transaction..." : "Submitting and waiting for confirmation...");
    try {
        const result = await submit(request);
        spinner.stop();
        displayResult(result);
    }
    catch (err) {
        spinner.stop();
        output.error(err.message);
        process.exit(1);
    }
}
