// @bankr/cli/dist/lib/api.js
import { CLI_USER_AGENT, getApiUrl, requireApiKey } from "./config.js";
const TERMINAL_STATUSES = [
    "completed",
    "failed",
    "cancelled",
];
function isTerminalStatus(status) {
    return TERMINAL_STATUSES.includes(status);
}
function authHeaders() {
    return {
        "X-API-Key": requireApiKey(),
        "Content-Type": "application/json",
        "User-Agent": CLI_USER_AGENT,
    };
}
async function handleResponse(res) {
    if (!res.ok) {
        const body = (await res
            .json()
            .catch(() => ({ message: res.statusText })));
        const msg = body.message || body.error || res.statusText;
        throw new Error(`API error (${res.status}): ${msg}`);
    }
    return res.json();
}
export async function submitPrompt(prompt, threadId) {
    const res = await fetch(`${getApiUrl()}/agent/prompt`, {
        method: "POST",
        headers: authHeaders(),
        body: JSON.stringify({ prompt, ...(threadId && { threadId }) }),
    });
    return handleResponse(res);
}
export async function getJobStatus(jobId) {
    const res = await fetch(`${getApiUrl()}/agent/job/${jobId}`, {
        headers: authHeaders(),
    });
    return handleResponse(res);
}
export async function cancelJob(jobId) {
    const res = await fetch(`${getApiUrl()}/agent/job/${jobId}/cancel`, {
        method: "POST",
        headers: authHeaders(),
    });
    return handleResponse(res);
}
export async function validateApiKey() {
    try {
        const res = await fetch(`${getApiUrl()}/_health`, {
            headers: { "X-API-Key": requireApiKey(), "User-Agent": CLI_USER_AGENT },
        });
        return res.ok;
    }
    catch {
        return false;
    }
}
export async function getUserInfo() {
    const res = await fetch(`${getApiUrl()}/agent/me`, {
        headers: authHeaders(),
    });
    return handleResponse(res);
}
export async function pollJob(jobId, opts = {}) {
    const { interval = 2000, maxAttempts = 150, onStatus } = opts;
    let attempts = 0;
    while (attempts < maxAttempts) {
        const status = await getJobStatus(jobId);
        onStatus?.(status);
        if (isTerminalStatus(status.status)) {
            return status;
        }
        await new Promise((resolve) => setTimeout(resolve, interval));
        attempts++;
    }
    throw new Error(`Polling timed out after ${maxAttempts} attempts`);
}
export async function sign(request) {
    const res = await fetch(`${getApiUrl()}/agent/sign`, {
        method: "POST",
        headers: authHeaders(),
        body: JSON.stringify(request),
    });
    return handleResponse(res);
}
export async function submit(request) {
    const res = await fetch(`${getApiUrl()}/agent/submit`, {
        method: "POST",
        headers: authHeaders(),
        body: JSON.stringify(request),
    });
    return handleResponse(res);
}
