// @bankr/cli/dist/lib/config.js
import { existsSync, mkdirSync, readFileSync, unlinkSync, writeFileSync, } from "fs";
import { homedir } from "os";
import { dirname, join } from "path";
let configPathOverride;
const DEFAULT_CONFIG_FILE = join(homedir(), ".bankr", "config.json");
export function setConfigPath(path) {
    configPathOverride = path;
}
export function getConfigPath() {
    return configPathOverride || process.env.BANKR_CONFIG || DEFAULT_CONFIG_FILE;
}
export const DEFAULT_API_URL = "https://api.bankr.bot";
export const DEFAULT_LLM_URL = "https://llm.bankr.bot";
export const CLI_USER_AGENT = "bankr-cli/0.1";
function ensureConfigDir() {
    const configDir = dirname(getConfigPath());
    if (!existsSync(configDir)) {
        mkdirSync(configDir, { recursive: true, mode: 0o700 });
    }
}
export function readConfig() {
    const configFile = getConfigPath();
    if (!existsSync(configFile)) {
        return {};
    }
    const raw = readFileSync(configFile, "utf-8");
    try {
        return JSON.parse(raw);
    }
    catch {
        console.warn(`Warning: Could not parse ${configFile}. Using defaults.`);
        return {};
    }
}
export function writeConfig(config) {
    ensureConfigDir();
    writeFileSync(getConfigPath(), JSON.stringify(config, null, 2) + "\n", {
        encoding: "utf-8",
        mode: 0o600,
    });
}
export function clearConfig() {
    const configFile = getConfigPath();
    if (existsSync(configFile)) {
        unlinkSync(configFile);
    }
}
export function getApiKey() {
    return process.env.BANKR_API_KEY || readConfig().apiKey;
}
export function getLlmKey() {
    return process.env.BANKR_LLM_KEY || readConfig().llmKey || getApiKey();
}
export function getApiUrl() {
    return process.env.BANKR_API_URL || readConfig().apiUrl || DEFAULT_API_URL;
}
export function getLlmUrl() {
    return process.env.BANKR_LLM_URL || readConfig().llmUrl || DEFAULT_LLM_URL;
}
export function isCustomApiUrl() {
    return getApiUrl() !== DEFAULT_API_URL;
}
export function getLastThreadId() {
    return readConfig().lastThreadId;
}
export function setLastThreadId(id) {
    const config = readConfig();
    config.lastThreadId = id;
    writeConfig(config);
}
export function requireApiKey() {
    const key = getApiKey();
    if (!key) {
        throw new Error("Not authenticated. Run `bankr login` to set up your Bankr API key.");
    }
    return key;
}
