// @bankr/cli/dist/lib/output.js
import chalk from "chalk";
import ora from "ora";
const BRAND_HEX = "#FF613D";
const brandColor = chalk.hex(BRAND_HEX);
const brandColorBold = brandColor.bold;
/** String formatters (return colored strings without logging). */
export const fmt = {
    brand: brandColor,
    brandBold: brandColorBold,
    dim: chalk.dim,
};
export const bankrTheme = {
    prefix: brandColor("▸"),
    style: {
        highlight: brandColorBold,
        answer: brandColor,
    },
    icon: {
        cursor: brandColor("●"),
    },
};
export function brand(msg) {
    console.log(brandColor(msg));
}
export function brandBold(msg) {
    console.log(brandColorBold(msg));
}
export function success(msg) {
    console.log(chalk.green(`✓ ${msg}`));
}
export function error(msg) {
    console.error(chalk.red(`✗ ${msg}`));
}
export function warn(msg) {
    console.log(chalk.yellow(`⚠ ${msg}`));
}
export function info(msg) {
    console.log(brandColor(msg));
}
export function blank() {
    console.log();
}
export function dim(msg) {
    console.log(chalk.dim(msg));
}
export function muted(msg) {
    console.log(chalk.hex("#666666")(msg));
}
export function label(key, value) {
    console.log(`${brandColorBold(key + ":")}  ${value}`);
}
export function keyValue(key, value) {
    const padding = Math.max(0, 18 - key.length);
    console.log(`  ${chalk.dim(key + ":")}${" ".repeat(padding)} ${value}`);
}
export function spinner(text) {
    return ora({ text, spinner: "dots", color: "yellow" }).start();
}
export function maskApiKey(key) {
    if (key.length <= 8)
        return "****";
    return key.slice(0, 6) + "..." + key.slice(-4);
}
export function formatDuration(ms) {
    if (ms < 1000)
        return `${ms}ms`;
    return `${(ms / 1000).toFixed(1)}s`;
}
export function formatStatus(status) {
    switch (status) {
        case "completed":
            return chalk.green(status);
        case "failed":
            return chalk.red(status);
        case "cancelled":
            return chalk.yellow(status);
        case "processing":
            return brandColor(status);
        case "pending":
            return chalk.dim(status);
        default:
            return status;
    }
}
